MBS-8145: Treat timeouts differently from ISEs

uklauer opened the pull request:
Database timeouts used to be treated as an internal server error, i.e. an email was sent to MetaBrainz employees about each timeout, and a rather technical error page with a stack trace shown to
the user. When, on the other hand, a request exceeded the allowable time, the server process was
immediately killed with a notice in the logs. The user was shown a plain nginx error page.

This, instead, suppresses error emails for both kinds of timeouts (but logs them) and shows a dedicated error page suggesting to reload the page, with a 503 status code (service unavailable). Instead of killing the process, an exception is thrown, allowing the normal error processing and cleanup to happen. This also makes the return-to-search-form feature of the edit search less fragile (it only worked when the query timed out before the request did, MBS-8382).

uklauer commented:
I tested this on a local server where I added `sleep;`, `die "foo";` and `...::StatementTimedOut->throw(...);` in a controller method to cause a request timeout and to simulate ISE-worthy errors and database timeouts, respectively.

uklauer commented on root/main/timeout.tt:
```diff
@@ -0,0 +1,8 @@
+
+        <p><strong>[% l('Processing your request took too long and timed out.') %]</strong></p>
+        <p>[% l('It may help to try again by reloading the page.') %]</p>
+    </div>
```
Feel free to comment on the language, of course.

I was also wondering whether we should include the strack trace here, perhaps in an element that is `display: none`? It may still be useful for developers without access to the log.

mwiencek replied:
Yeah, that sounds like a good idea. The message otherwise reads fine to me.

mwiencek commented on lib/MusicBrainz/Server.pm:
```diff
@@ -367,19 +367,11 @@ around dispatch => sub {
                 $sth->cancel;
             }
-
-            $context->connector->disconnect;
```
No longer needed because it'll make it to the end of `dispatch` now, right?

uklauer replied:
Yes, exactly.

mwiencek commented:
This is wonderful. :) We can merge after the hidden stack trace is added, if you'd like to do that.

uklauer commented:
I added a `<div style="display:none">` to the timeout page containing the information that is shown for ISEs. To this end, I took the relevant code blocks from `500.tt` and put them into two new files, `error.tt` with the error string and stack trace and `environment.tt` with things such as time and processing server.

In the process, I moved the request parameters from `error.tt` to `environment.tt` because it seemed more fitting there, and also made `400.tt` use these blocks (if 400s outside of the WS even occur).

uklauer commented on root/main/400.tt:
```diff
@@ -4,15 +4,7 @@
-            [% END %]
-        </p>
-
+        [% INCLUDE 'main/info/error.tt' %]
```
When there is no stack trace in the stash, `error.tt` resolves exactly to the above.

gentlecat commented:
Is there a way to keep track of number of timeouts that we have apart from looking at the logs directly? Something in http://stats.musicbrainz.org/?

mwiencek commented:
@gentlecat Hmm, not really. You can see the number of 503s on stats.mb, but those will almost all be from rate-limiting. We at least know exactly what pages tend to time out (or it's obvious while using the site), so the emails weren't all that helpful (and it's easy to run `grep -R 'Request took more than' .` in the service directory if we need examples).
